### PR TITLE
feature: support update hostname

### LINKED ISSUE #2179: [proposal] pouch supports to update hostname for rich container
### Ⅰ. Issue Description
When creating a container, pouch can set container hostname by `--hostname` option. Once hostname was set, It couldn't be updated in container config. It will be nice that pouch can update hostname for rich container and also its config. Maybe `pouch update` supporting `--hostname` for a running rich container  is reasonable.

### Ⅳ. How to reproduce it (as minimally and precisely as possible)

1.
2.
3.

### Ⅵ. Environment:

- pouch version (use `pouch version`):
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by wangforthinker**
Signed-off-by: allen.wang <allen.wq@alipay.com>

### Ⅰ. Describe what this PR did
Consider to update container's hostname in a running container, It should join container namespace. So it imports package github.com/opencontainers/runc/libcontainer/nsenter to meet the requirement(about more implementation details refferring to Ⅳ. Describe how to verify it). Otherwise, It provides more register functions which can do more in container namespace. 

### Ⅱ. Does this pull request fix one issue?

fixes #2179 

### Ⅲ. Why don't you add test cases (unit test/integration test)? (你真的觉得不需要加测试吗？)
add test TestRunningContainerUpdateHostname and TestStoppedContainerUpdateHostname.

### Ⅳ. Describe how to verify it
after doing update hostname, we can exec into container, and run command "hostname" to verify it.  

### Ⅴ. Special notes for reviews
It adds a package pkg/nsexec, which defines a function NsExec to join container namespace.  It provides register functions map to allow do something in container namespace such as update hostname.  It registers a reexec command called nsexec, which imports  github.com/opencontainers/runc/libcontainer/nsenter to join container namespace by calling C code. Package pkg/nsexec executes new command to call pouchd nsexec.

### CODE REVIEW

**Comment by allencloud** [comment]
I think this pr has updated or added too many codes, so I am afraid we have to add more pr description s to make the demand and implementation clearly. @wangforthinker 
Thanks a lot. 😄

File: pkg/nsexec/nsexec_linux.go [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+	cmd.Env = append(cmd.Env,
+		fmt.Sprintf("_LIBCONTAINER_INITPIPE=%d", stdioFdCount+len(cmd.ExtraFiles)-1),
+	)
+	cmd.Stdin = p.stdin
+	cmd.Stdout = p.stdout
+	cmd.Stderr = p.stderr
+	p.cmd = cmd
+}
+
+//setns is to wait init process, because nsenter will
```
Comment (Ace-Tang) [inline pkg/nsexec/nsexec_linux.go]: I think this comment is not finished.

Comment (wangforthinker) [reply pkg/nsexec/nsexec_linux.go]: I have updated it. @Ace-Tang .

File: daemon/mgr/container.go [hunk 2]
```diff
@@ -941,6 +941,19 @@ func (mgr *ContainerManager) Update(ctx context.Context, name string, config *ty
 		}
 	}()
 
+	if config.Hostname != "" {
+		if c.HostConfig.NetworkMode == "host" {
+			return fmt.Errorf("host container %s cannot update hostname", c.ID)
```
Comment (HusterWan) [inline daemon/mgr/container.go]: how about `cannot update hostname of container %s  in host network mode `  ?

File: daemon/mgr/container.go [hunk 3]
```diff
@@ -941,6 +941,19 @@ func (mgr *ContainerManager) Update(ctx context.Context, name string, config *ty
 		}
 	}()
 
+	if config.Hostname != "" {
```
Comment (HusterWan) [inline daemon/mgr/container.go]: Just asking, why we cannot update hostname for container not set before?

File: main.go [hunk 4]
```diff
@@ -35,9 +35,14 @@ var (
...
 )
 
+func init() {
+	reexec.Register("nsexec", reexecRunNsExec)
+}
+
 var cfg = &config.Config{}
 
 func main() {
+	logrus.Infof("os args:%v", os.Args)
```
Comment (HusterWan) [inline main.go]: No need log here? WDYT?

File: nsexec.go [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+import (
+	"encoding/json"
+	"os"
+	"strconv"
+
+	_ "github.com/opencontainers/runc/libcontainer/nsenter"
+	"github.com/sirupsen/logrus"
+
+	"github.com/alibaba/pouch/pkg/nsexec"
```
Comment (HusterWan) [inline nsexec.go]: third party package should put in third part.

File: nsexec.go [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+)
+
+//reexecRunNsExec will run nsenter in go init and join to container namespace
```
Comment (HusterWan) [inline nsexec.go]: put a space between `//` and `reexecRunNsExec`

File: pkg/nsexec/nsexec_linux.go [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+func newExecProcess(stdin io.Reader, stdout io.Writer, stderr io.Writer) *process {
+	return &process{
+		initArgs: []string{"/proc/self/exe", "nsexec"},
+		stdin:    stdin,
+		stdout:   stdout,
+		stderr:   stderr,
+	}
+}
+
+func newNsExecCmd(p *process) {
```
Comment (HusterWan) [inline pkg/nsexec/nsexec_linux.go]: can we add test cases for this package?

File: daemon/mgr/container.go [hunk 8]
```diff
@@ -941,6 +941,19 @@ func (mgr *ContainerManager) Update(ctx context.Context, name string, config *ty
...
+		if c.HostConfig.NetworkMode == "host" {
+			return fmt.Errorf("host container %s cannot update hostname", c.ID)
+		}
+
+		c.Config.Hostname = strfmt.Hostname(config.Hostname)
```
Comment (HusterWan) [inline daemon/mgr/container.go]: Should we add a validation for the hostname that user input?

File: daemon/mgr/container.go [hunk 9]
```diff
@@ -941,6 +941,19 @@ func (mgr *ContainerManager) Update(ctx context.Context, name string, config *ty
...
+		if c.IsRunning() {
+			if err = updateHostnameForRunningContainer(config.Hostname, c); err != nil {
+				return err
```
Comment (HusterWan) [inline daemon/mgr/container.go]: How about adding a log  to record when updating the hostname of container successful?

File: daemon/mgr/container_update_hostname.go [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+
+	err = syscall.Sethostname(data)
+	if err != nil {
+		return nil, err
+	}
+
+	//update /etc/hostname and /etc/hosts
+	err = ioutil.WriteFile("/etc/hostname", append(data, '\n'), 0644)
+	if err != nil {
+		return nil, fmt.Errorf("update /etc/hostname failed:%s", err.Error())
```
Comment (HusterWan) [inline daemon/mgr/container_update_hostname.go]: How about `fmt.Errorf("failed to update /etc/hostname: %v", err) ?

File: daemon/mgr/container.go (hunk 3 above)
Comment (wangforthinker) [reply daemon/mgr/container.go]: This config is updateConfig.If updateConfig.Hostname is nil, it means hostname not need to be updated. @HusterWan

File: daemon/mgr/container.go (hunk 9 above)
Comment (wangforthinker) [reply daemon/mgr/container.go]: Have added a log to record it .

File: daemon/mgr/container_update_hostname.go [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+	d := string(hosts)
+	lines := strings.Split(d, "\n")
+	for i, line := range lines {
+		if !strings.Contains(line, oldHostname) {
+			continue
+		}
+
+		newLine := line
+
+		if exp1.FindString(newLine) != "" {
```
[the remark this answers is not in the document]
Comment (wangforthinker) [reply daemon/mgr/container_update_hostname.go]: One line may have all two model.

File: pkg/nsexec/nsexec_linux.go (hunk 7 above)
Comment (wangforthinker) [reply pkg/nsexec/nsexec_linux.go]: The main function NsExec calls a cli command(/proc/self/exe nsexec), it is difficult to add go test for it! We may only test it by cli test.

File: daemon/mgr/container_update_hostname.go [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+func updateHostnameForRunningContainer(hostname string, container *Container) error {
+	pid := container.State.Pid
+	if pid <= 0 {
+		return fmt.Errorf("container pid %d is not vaild", pid)
+	}
+
+	_, err := nsexec.NsExec(int(pid), nsexec.Op{OpCode: "sethostname", Data: []byte(hostname)})
+	if err != nil {
```
Comment (HusterWan) [inline daemon/mgr/container_update_hostname.go]: direct return err?

File: daemon/mgr/network_utils.go [hunk 13]
```diff
@@ -9,6 +9,11 @@ import (
 	"github.com/docker/libnetwork"
 )
 
+const (
+	// NetworkModeHost means container network mode is host
+	NetworkModeHost = "host"
```
Comment (HusterWan) [inline daemon/mgr/network_utils.go]: can we add more network mode: none, bridge, default?

File: daemon/mgr/spec_linux.go [hunk 14]
```diff
@@ -387,7 +387,7 @@ func setupNamespaces(ctx context.Context, c *Container, specWrapper *SpecWrapper
 
 // isHost indicates whether the container shares the host's corresponding namespace.
 func isHost(mode string) bool {
```
Comment (HusterWan) [inline daemon/mgr/spec_linux.go]: duplicate with `IsHost` ?

**Pull request closed without merging**